handler/src/main/java/io/netty/handler/ssl/ResumptionController.java

```diff
@@ -0,0 +1,169 @@
+        if (resumableTm.get() != null) {
+            confirmedValidations.remove(unwrapEngine(engine));
+        }
+    }
+
+    public boolean validateResumeIfNeeded(SSLEngine engine)
+            throws CertificateException, SSLPeerUnverifiedException {
+        ResumableX509ExtendedTrustManager tm;
+        boolean valid = engine.getSession().isValid();
+        if (valid && (tm = resumableTm.get()) != null) {
+            Certificate[] peerCertificates = engine.getSession().getPeerCertificates();
+            engine = unwrapEngine(engine);
+            if (confirmedValidations.remove(engine) == null) {
+                // This is a resumed session.
+                if (engine.getUseClientMode()) {
+                    // We are the client, resuming a session trusting the server
+                    tm.resumeServerTrusted(chainOf(peerCertificates), engine);
+                } else {
+                    // We are the server, resuming a session trusting the client
+                    tm.resumeClientTrusted(chainOf(peerCertificates), engine);
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static SSLEngine unwrapEngine(SSLEngine engine) {
+        if (engine instanceof JdkSslEngine) {
+            return ((JdkSslEngine) engine).getWrappedEngine();
+        }
+        return engine;
+    }
+
+    private static X509Certificate[] chainOf(Certificate[] peerCertificates) {
+        if (peerCertificates instanceof X509Certificate[]) {
+            //noinspection SuspiciousArrayCast
+            return (X509Certificate[]) peerCertificates;
+        }
+        X509Certificate[] chain = new X509Certificate[peerCertificates.length];
+        for (int i = 0; i < peerCertificates.length; i++) {
+            Certificate cert = peerCertificates[i];
+            if (cert instanceof X509Certificate) {
+                chain[i] = (X509Certificate) cert;
+            } else {
+                throw new IllegalArgumentException("Only X509Certificates are supported, found: " +
+                        (cert == null ? null : cert.getClass()));
+            }
+        }
+        return chain;
+    }
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    private static final class X509ExtendedWrapTrustManager extends X509ExtendedTrustManager {
+        private final X509ExtendedTrustManager trustManager;
+        private final Map<SSLEngine, Object> confirmedValidations;
+
+        X509ExtendedWrapTrustManager(X509ExtendedTrustManager trustManager,
+                                     Map<SSLEngine, Object> confirmedValidations) {
+            this.trustManager = ObjectUtil.checkNotNull(trustManager, "trustManager");
+            this.confirmedValidations = confirmedValidations;
+        }
+
+        private static void unsupported() throws CertificateException {
+            throw new CertificateException("Resumable trust managers require the SSLEngine parameter");
```

Comment:
Why not throw `UnsupportedOperationException` ?

Reply:
hmm this is kind of stupid but ok... Maybe just wrap an `UnsupportedOperationException` in it then...